File: src/core/sync.rs
// src/core/sync.rs
use std::collections::HashSet;
use std::fs;
use std::path::Path;
use anyhow::Result;
use crate::{defs, utils, core::inventory::Module};

/// Synchronizes enabled modules from source to the prepared storage.
/// Implements a smart sync strategy to avoid unnecessary I/O.
pub fn perform_sync(modules: &[Module], target_base: &Path) -> Result<()> {
    log::info!("Starting smart module sync to {}", target_base.display());

    // 1. Prune orphaned directories (Cleanup disabled/removed modules)
    prune_orphaned_modules(modules, target_base)?;

    // 2. Sync each module
    for module in modules {
        let dst = target_base.join(&module.id);
        
        // Recursively check if the module has actual content for known partitions
        let has_content = defs::BUILTIN_PARTITIONS.iter().any(|p| {
            let part_path = module.source_path.join(p);
            part_path.exists() && has_files_recursive(&part_path)
        });

        if has_content {
            if should_sync(&module.source_path, &dst) {
                log::info!("Syncing module: {} (Updated/New)", module.id);
                
                // Ensure clean state for this module before copying
                if dst.exists() {
                    if let Err(e) = fs::remove_dir_all(&dst) {
                        log::warn!("Failed to clean target dir for {}: {}", module.id, e);
                    }
                }

                if let Err(e) = utils::sync_dir(&module.source_path, &dst) {
                    log::error!("Failed to sync module {}: {}", module.id, e);
                } else {
                    // 3. Context Mirroring (Only needed after a fresh sync)
                    repair_module_contexts(&dst, &module.id);
                }
            } else {
                log::debug!("Skipping module: {} (Up-to-date)", module.id);
            }
        } else {
            log::debug!("Skipping empty module: {}", module.id);
        }
    }
    
    Ok(())
}

/// Removes directories in the target base that do not correspond to any active module.
fn prune_orphaned_modules(modules: &[Module], target_base: &Path) -> Result<()> {
    if !target_base.exists() { return Ok(()); }

    let active_ids: HashSet<&str> = modules.iter().map(|m| m.id.as_str()).collect();

    for entry in fs::read_dir(target_base)? {
        let entry = entry?;
        let path = entry.path();
        let name_os = entry.file_name();
        let name = name_os.to_string_lossy();

        // Skip internal files/dirs
        if name == "lost+found" || name == "meta-hybrid" { continue; }

        if !active_ids.contains(name.as_ref()) {
            log::info!("Pruning orphaned module storage: {}", name);
            if path.is_dir() {
                if let Err(e) = fs::remove_dir_all(&path) {
                    log::warn!("Failed to remove orphan dir {}: {}", name, e);
                }
            } else {
                if let Err(e) = fs::remove_file(&path) {
                    log::warn!("Failed to remove orphan file {}: {}", name, e);
                }
            }
        }
    }
    Ok(())
}

/// Determines if a module needs to be synced.
/// Compares `module.prop` content as a heuristic for version/content changes.
fn should_sync(src: &Path, dst: &Path) -> bool {
    if !dst.exists() {
        return true;
    }

    // Compare module.prop
    let src_prop = src.join("module.prop");
    let dst_prop = dst.join("module.prop");

    if !src_prop.exists() || !dst_prop.exists() {
        // If prop file is missing in either, force sync to be safe
        return true;
    }

    // Read and compare contents
    // We use read_to_string/bytes. If checking file size/mtime is preferred for speed,
    // we could do that, but content check is more robust against 'touch'.
    match (fs::read(&src_prop), fs::read(&dst_prop)) {
        (Ok(s), Ok(d)) => s != d, // Sync if content differs
        _ => true, // Sync on IO errors
    }
}

fn repair_module_contexts(module_root: &Path, module_id: &str) {
    for part in defs::BUILTIN_PARTITIONS {
        let part_root = module_root.join(part);
        if part_root.exists() {
            if let Err(e) = recursive_context_repair(module_root, &part_root) {
                log::warn!("Context repair failed for {}/{}: {}", module_id, part, e);
            }
        }
    }
}

fn recursive_context_repair(base: &Path, current: &Path) -> Result<()> {
    if !current.exists() { return Ok(()); }
    
    // Calculate path relative to module root to find system equivalent
    // e.g. /mnt/modA/system/bin/app -> /system/bin/app
    let relative = current.strip_prefix(base)?;
    let system_path = Path::new("/").join(relative);

    if system_path.exists() {
        // Copy context from real system file
        let _ = utils::copy_path_context(&system_path, current);
    }

    if current.is_dir() {
        for entry in fs::read_dir(current)? {
            let entry = entry?;
            recursive_context_repair(base, &entry.path())?;
        }
    }
    Ok(())
}

fn has_files_recursive(path: &Path) -> bool {
    if let Ok(entries) = fs::read_dir(path) {
        for entry in entries.flatten() {
            if let Ok(ft) = entry.file_type() {
                if ft.is_dir() {
                    if has_files_recursive(&entry.path()) { return true; }
                } else {
                    return true; // Found a file/symlink/device
                }
            }
        }
    }
    false
}

File: src/core/sync.cpp
// core/sync.cpp - Module content sync
#include "sync.hpp"
#include <fstream>
#include <set>
#include "../defs.hpp"
#include "../utils.hpp"

namespace hymo {

// Check if module has content for any partition
static bool has_content(const fs::path& module_path,
                        const std::vector<std::string>& all_partitions) {
    for (const auto& partition : all_partitions) {
        fs::path part_path = module_path / partition;
        if (has_files_recursive(part_path)) {
            return true;
        }
    }
    return false;
}

// Check if module needs sync by comparing module.prop
static bool should_sync(const fs::path& src, const fs::path& dst) {
    if (!fs::exists(dst)) {
        return true;  // New module
    }

    fs::path src_prop = src / "module.prop";
    fs::path dst_prop = dst / "module.prop";

    if (!fs::exists(src_prop) || !fs::exists(dst_prop)) {
        return true;  // Force sync if prop missing
    }

    try {
        std::ifstream src_file(src_prop, std::ios::binary);
        std::ifstream dst_file(dst_prop, std::ios::binary);

        std::string src_content((std::istreambuf_iterator<char>(src_file)),
                                std::istreambuf_iterator<char>());
        std::string dst_content((std::istreambuf_iterator<char>(dst_file)),
                                std::istreambuf_iterator<char>());

        return src_content != dst_content;
    } catch (...) {
        return true;
    }
}

// Remove orphaned module directories
static void prune_orphaned_modules(const std::vector<Module>& modules,
                                   const fs::path& storage_root) {
    if (!fs::exists(storage_root)) {
        return;
    }

    std::set<std::string> active_ids;
    for (const auto& module : modules) {
        active_ids.insert(module.id);
    }

    try {
        for (const auto& entry : fs::directory_iterator(storage_root)) {
            std::string name = entry.path().filename().string();

            if (name == "lost+found" || name == "hymo") {
                continue;
            }

            if (active_ids.find(name) == active_ids.end()) {
                LOG_INFO("Pruning orphaned storage: " + name);
                try {
                    fs::remove_all(entry.path());
                } catch (const std::exception& e) {
                    LOG_WARN("Failed to remove: " + name);
                }
            }
        }
    } catch (...) {
        LOG_WARN("Failed to prune orphans.");
    }
}

// Map SELinux context from system if possible
static void recursive_context_repair(const fs::path& base, const fs::path& current) {
    if (!fs::exists(current)) {
        return;
    }

    try {
        std::string file_name = current.filename().string();

        // Use parent context for internal overlay structs
        if (file_name == "upperdir" || file_name == "workdir") {
            if (current.has_parent_path()) {
                fs::path parent = current.parent_path();
                try {
                    std::string parent_ctx = lgetfilecon(parent);
                    lsetfilecon(current, parent_ctx);
                } catch (...) {
                }
            }
        } else {
            fs::path relative = fs::relative(current, base);
            fs::path system_path = fs::path("/") / relative;

            if (fs::exists(system_path)) {
                copy_path_context(system_path, current);
            }
        }

        if (fs::is_directory(current)) {
            for (const auto& entry : fs::directory_iterator(current)) {
                recursive_context_repair(base, entry.path());
            }
        }
    } catch (const std::exception& e) {
        LOG_DEBUG("Context repair failed: " + current.string());
    }
}

static void repair_module_contexts(const fs::path& module_root, const std::string& module_id,
                                   const std::vector<std::string>& all_partitions) {
    LOG_DEBUG("Repairing SELinux contexts for: " + module_id);

    for (const auto& partition : all_partitions) {
        fs::path part_root = module_root / partition;

        if (fs::exists(part_root) && fs::is_directory(part_root)) {
            try {
                recursive_context_repair(module_root, part_root);
            } catch (const std::exception& e) {
                LOG_WARN("Context repair failed for " + module_id + "/" + partition);
            }
        }
    }
}

void perform_sync(const std::vector<Module>& modules, const fs::path& storage_root,
                  const Config& config) {
    LOG_INFO("Syncing modules to " + storage_root.string());

    std::vector<std::string> all_partitions = BUILTIN_PARTITIONS;
    for (const auto& part : config.partitions) {
        all_partitions.push_back(part);
    }

    prune_orphaned_modules(modules, storage_root);

    for (const auto& module : modules) {
        fs::path dst = storage_root / module.id;

        if (!has_content(module.source_path, all_partitions)) {
            LOG_DEBUG("Skipping empty module: " + module.id);
            continue;
        }

        if (should_sync(module.source_path, dst)) {
            LOG_DEBUG("Syncing: " + module.id);

            if (fs::exists(dst)) {
                try {
                    fs::remove_all(dst);
                } catch (const std::exception& e) {
                    LOG_WARN("Failed to clean " + module.id);
                }
            }

            if (!sync_dir(module.source_path, dst)) {
                LOG_ERROR("Failed to sync: " + module.id);
            } else {
                repair_module_contexts(dst, module.id, all_partitions);
            }
        } else {
            LOG_DEBUG("Up-to-date: " + module.id);
        }
    }

    LOG_INFO("Sync completed.");
}

}  // namespace hymo